File: js/Chess/ChessProblem.js
import BoardParser from './BoardParser'
import PieceColor from './PieceColor'
import PieceType from './PieceType'
import ChessStatus from './ChessStatus'

export default class ChessProblem {
    static chessStatus;

    static loadFrom(lines) {
        this.board = new BoardParser().parseBoard(lines);
    }

    // Определяет мат, шах или пат белым.
    static calculateChessStatus() {
        const isCheck = ChessProblem.isCheckForWhite();
        let hasMoves = false;

        for (const locFrom of this.board.getPieces(PieceColor.white)) {
            for (const locTo of this.board.getPiece(locFrom).getMoves(locFrom, this.board)) {
                const old = this.board.getPiece(locTo);
                this.board.set(locTo, this.board.getPiece(locFrom));
                this.board.set(locFrom, null);
                if (!ChessProblem.isCheckForWhite())
                    hasMoves = true;
                this.board.set(locFrom, this.board.getPiece(locTo));
                this.board.set(locTo, old);
            }
        }

        if (isCheck)
            if (hasMoves)
                this.chessStatus = ChessStatus.check; // шах
            else this.chessStatus = ChessStatus.mate; // мат
        else if (hasMoves) this.chessStatus = ChessStatus.ok;
        else this.chessStatus = ChessStatus.stalemate; // пат
    }

    // check — это шах
    static isCheckForWhite() {
        let isCheck = false;
        for (const loc of this.board.getPieces(PieceColor.black)) {
            const piece = this.board.getPiece(loc);
            const moves = piece.getMoves(loc, this.board);
            for (const destination of moves) {
                const destinationPiece = this.board.getPiece(destination);
                if (destinationPiece && destinationPiece.is(PieceColor.white, PieceType.King))
                    isCheck = true;
            }
        }
        if (isCheck) return true;
        return false;
    }
}

File: python/Chess/chess_problem.py
from board_parser import BoardParser
from piece_color import PieceColor
from chess_status import ChessStatus
import piece_type


class ChessProblem:
    board = None
    chess_status = None

    @staticmethod
    def load_from(lines):
        ChessProblem.board = BoardParser().parse_board(lines)

    # Определяет мат, шах или пат белым.
    @staticmethod
    def calculate_chess_status():
        is_check = ChessProblem.is_check_for_white()
        has_movies = False

        for loc_from in ChessProblem.board.get_pieces(PieceColor.white):
            for loc_to in ChessProblem.board.get_piece(loc_from).get_moves(
                loc_from, ChessProblem.board
            ):
                old = ChessProblem.board.get_piece(loc_to)
                ChessProblem.board.set(loc_to, ChessProblem.board.get_piece(loc_from))
                ChessProblem.board.set(loc_from, None)
                if not ChessProblem.is_check_for_white():
                    has_movies = True
                ChessProblem.board.set(loc_from, ChessProblem.board.get_piece(loc_to))
                ChessProblem.board.set(loc_to, old)

        if is_check:
            if has_movies:
                ChessProblem.chess_status = ChessStatus.check  # шах
            else:
                ChessProblem.chess_status = ChessStatus.mate  # мат
        elif has_movies:
            ChessProblem.chess_status = ChessStatus.ok
        else:
            ChessProblem.chess_status = ChessStatus.stalemate  # пат

    # check — это шах
    @staticmethod
    def is_check_for_white():
        is_check = False
        for loc in ChessProblem.board.get_pieces(PieceColor.black):
            piece = ChessProblem.board.get_piece(loc)
            moves = piece.get_moves(loc, ChessProblem.board)
            for destination in moves:
                destination_piece = ChessProblem.board.get_piece(destination)
                if destination_piece and destination_piece.is_equal(
                    PieceColor.white, piece_type.King
                ):
                    is_check = True
        if is_check:
            return True
        return False
